moon/controller_round1.py

```diff
@@ -132,17 +144,52 @@ def on_object_reached(self, data):
         ):
             x,y,z = self.xyz
-            print(self.sim_time, "app: Object %s[%d] reached at [%.1f,%.1f], queueing up 19 around" % (object_type, object_index, x, y))
-            heapq.heappush(self.volatile_queue, (self.sim_time.total_seconds(), [object_index, object_type, x, y]))
+            print(self.sim_time, "app: Object %s[%d] reached at [%.1f,%.1f]" % (object_type, object_index, x, y))
+
+            # covered_area = { idx: { layer: Polygon() } }
+            # https://www.researchgate.net/figure/Minimum-overlap-of-circles-covering-an-area_fig2_256607366
+            # Minimum overlap of circles covering an area; r= 3.46 (=2* r * cos(30)); step 60 degrees, 6x
+            queued_count = 0
+            covered_area = defaultdict(lambda: defaultdict(lambda: Polygon()))
+            for key, p in self.volatile_queue:
+                layer, idx, _, qx, qy = p
+                for l in range(layer, 3):
+                    covered_area[idx][l] = covered_area[idx][l].union(Point(qx, qy).buffer(2))
+
+            olap = covered_area[object_index][0].intersection(Point(x,y).buffer(2)).area
+            if olap < 0.99 * 4 * math.pi: # if overlap is less than 100%
+                heapq.heappush(self.volatile_queue, (olap, [0, object_index, object_type, x, y]))
+                queued_count += 1
+
             for i in range(6):
                 x_d,y_d = pol2cart(3.46, math.radians(i * 60))
-                heapq.heappush(self.volatile_queue, (3600 + self.sim_time.total_seconds(), [object_index, object_type, x+x_d, y+y_d]))
-                 # https://www.researchgate.net/figure/Minimum-overlap-of-circles-covering-an-area_fig2_256607366
-                 # Minimum overlap of circles covering an area; r= 3.46 (=2* r * cos(30)); step 60 degrees, 6x
+                olap = covered_area[object_index][1].intersection(Point(x+x_d,y+y_d).buffer(2)).area
+                if olap < 0.99 * 4 * math.pi:
+                    heapq.heappush(self.volatile_queue, (1000 + olap, [1, object_index, object_type, x+x_d, y+y_d]))
+                    queued_count += 1
+
+            # further out non-overlapping guesses have higher priority than closer guesses with half circle or more of overlap with existing guesses
             for i in range(12):
                 x_d,y_d = pol2cart(2*3.46 if i % 2 == 0 else 2*3, math.radians(i * 30))
-                heapq.heappush(self.volatile_queue, (7200 + self.sim_time.total_seconds(), [object_index, object_type, x+x_d, y+y_d]))
-
+                olap = covered_area[object_index][2].intersection(Point(x+x_d,y+y_d).buffer(2)).area
+                if olap < 0.99 * 4 * math.pi:
+                    heapq.heappush(self.volatile_queue, (1006 + olap, [2, object_index, object_type, x+x_d, y+y_d]))
+                    queued_count += 1
+
+            print(self.sim_time, "app: Queueing %d guesses" % (queued_count))
+            if self.debug:
+                print("=== matplotlib start ===")
```

Comment:
:) interesting - so you dump drawing code into the log file and then you debug it from it ... OK